File: javascript/test/sas.test.js
const { Sas } = require('../pkg/vodozemac.js');

const EXTRA_INFO = "extra_info";
const MESSAGE = "Test message"

describe('Sas', function() {
    it('should be created successfully', function() {
        let alice = new Sas();

        expect(alice.public_key).not.toBe("");
    });

    it('should allow us to establish a shared secret', function() {
        let alice = new Sas();
        let bob = new Sas();
        const bob_public_key = bob.public_key;

        bob = bob.diffie_hellman(alice.public_key);
        alice = alice.diffie_hellman(bob_public_key);
    });

    it('should allow us to generate common short auth strings', function() {
        let alice = new Sas();
        let bob = new Sas();
        const bob_public_key = bob.public_key;

        bob = bob.diffie_hellman(alice.public_key);
        alice = alice.diffie_hellman(bob_public_key);

        const alice_bytes = alice.bytes(EXTRA_INFO);
        const bob_bytes = bob.bytes(EXTRA_INFO);

        expect(alice_bytes.emoji_indices).toEqual(bob_bytes.emoji_indices);
        expect(alice_bytes.decimals).toEqual(bob_bytes.decimals);
    });

    it('should allow us to generate a message authentication code', function() {
        let alice = new Sas();
        let bob = new Sas();
        const bob_public_key = bob.public_key;

        bob = bob.diffie_hellman(alice.public_key);
        alice = alice.diffie_hellman(bob_public_key);

        const alice_mac = alice.calculate_mac(MESSAGE, EXTRA_INFO);
        const bob_mac = bob.calculate_mac(MESSAGE, EXTRA_INFO);

        expect(() => alice.verify_mac("", EXTRA_INFO, bob_mac)).toThrow();
        alice.verify_mac(MESSAGE, EXTRA_INFO, bob_mac);
        bob.verify_mac(MESSAGE, EXTRA_INFO, alice_mac);
        expect(alice_mac).toEqual(bob_mac);
    });
});

File: python/tests/sas_test.py
import pytest

from vodozemac import SasException, Sas

MESSAGE = "Test message"
EXTRA_INFO = "extra_info"


class TestClass(object):
    def test_sas_creation(self):
        sas = Sas()
        assert sas.public_key

    def test_other_key_setting(self):
        sas_alice = Sas()
        sas_bob = Sas()

        established = sas_alice.diffie_hellman(sas_bob.public_key)

    def test_bytes_generating(self):
        sas_alice = Sas()
        sas_bob = Sas()

        bob_public_key = sas_bob.public_key
        sas_bob = sas_bob.diffie_hellman(sas_alice.public_key)
        sas_alice = sas_alice.diffie_hellman(bob_public_key)

        alice_bytes = sas_alice.bytes(EXTRA_INFO)
        bob_bytes = sas_bob.bytes(EXTRA_INFO)

        assert alice_bytes.emoji_indices == bob_bytes.emoji_indices
        assert alice_bytes.decimals == bob_bytes.decimals

    def test_mac_generating(self):
        sas_alice = Sas()
        sas_bob = Sas()

        bob_public_key = sas_bob.public_key
        sas_bob = sas_bob.diffie_hellman(sas_alice.public_key)
        sas_alice = sas_alice.diffie_hellman(bob_public_key)

        alice_mac = sas_alice.calculate_mac(MESSAGE, EXTRA_INFO)
        bob_mac = sas_bob.calculate_mac(MESSAGE, EXTRA_INFO)

        sas_alice.verify_mac(MESSAGE, EXTRA_INFO, bob_mac)
        sas_bob.verify_mac(MESSAGE, EXTRA_INFO, alice_mac)

        assert alice_mac == bob_mac
